File: src/main/java/com/example/demo/ManifestClient.java
package com.example.demo;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.springframework.web.client.RestTemplate;

public class ManifestClient {
    private final RestTemplate restTemplate = new RestTemplate();
    private final ObjectMapper objectMapper = new ObjectMapper();

    private JsonNode manifest;
    
    public JsonNode js(String entry) {
      this.fetchManifest();
      if (manifest != null && manifest.has("entrypoints")) {
          JsonNode entryNode = manifest.get("entrypoints").get(entry);
          if (entryNode != null && entryNode.has("assets") && entryNode.get("assets").has("js")) {
              return entryNode.get("assets").get("js");
          }
      }

      return objectMapper.createArrayNode();
    }

    public JsonNode css(String entry) {
        this.fetchManifest();
        if (manifest != null && manifest.has("entrypoints")) {
            JsonNode entryNode = manifest.get("entrypoints").get(entry);

            if (entryNode != null && entryNode.has("assets") && entryNode.get("assets").has("css")) {
                return manifest.get("entrypoints").get(entry).get("assets").get("css");
            }
        }

        return objectMapper.createArrayNode();
    }

    private void fetchManifest() {
//        if (this.manifest != null) {
//            return;
//        }
        String url = "http://localhost:3013/version-dev/manifest.json";
        String manifestStr = restTemplate.getForObject(url, String.class);
        try {
            this.manifest = objectMapper.readTree(manifestStr);
        } catch (Exception e) {
            this.manifest = null;
        }
    }
}

File: internal/web/manifest_client.go
package web

import (
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"sync"
)

type ManifestClient struct {
	ManifestHost string
	manifest     map[string]interface{}
	mu           sync.Mutex
}

func NewManifestClient(manifestHost string) *ManifestClient {
	return &ManifestClient{ManifestHost: manifestHost}
}

func (mc *ManifestClient) fetchManifest() error {
	mc.mu.Lock()
	defer mc.mu.Unlock()

	// TODO: memoize manifest in production
	//if mc.manifest != nil {
	//    return nil
	//}

	url := fmt.Sprintf("%s/version-dev/manifest.json", mc.ManifestHost)
	resp, err := http.Get(url)
	if err != nil {
		mc.manifest = nil
		return err
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		mc.manifest = nil
		return err
	}

	var manifest map[string]interface{}
	if err := json.Unmarshal(body, &manifest); err != nil {
		mc.manifest = nil
		return err
	}
	mc.manifest = manifest
	return nil
}

func (mc *ManifestClient) JS(entry string) []string {
	mc.fetchManifest()
	entrypoints, ok := mc.manifest["entrypoints"].(map[string]interface{})
	if !ok {
		return []string{}
	}
	entryNode, ok := entrypoints[entry].(map[string]interface{})
	if !ok {
		return []string{}
	}
	assets, ok := entryNode["assets"].(map[string]interface{})
	if !ok {
		return []string{}
	}
	jsFiles, ok := assets["js"].([]interface{})
	if !ok {
		return []string{}
	}
	result := make([]string, 0, len(jsFiles))
	for _, v := range jsFiles {
		if s, ok := v.(string); ok {
			result = append(result, s)
		}
	}
	return result
}

func (mc *ManifestClient) CSS(entry string) []string {
	mc.fetchManifest()
	entrypoints, ok := mc.manifest["entrypoints"].(map[string]interface{})
	if !ok {
		return []string{}
	}
	entryNode, ok := entrypoints[entry].(map[string]interface{})
	if !ok {
		return []string{}
	}
	assets, ok := entryNode["assets"].(map[string]interface{})
	if !ok {
		return []string{}
	}
	cssFiles, ok := assets["css"].([]interface{})
	if !ok {
		return []string{}
	}
	result := make([]string, 0, len(cssFiles))
	for _, v := range cssFiles {
		if s, ok := v.(string); ok {
			result = append(result, s)
		}
	}
	return result
}
